File: VN/VNData/lua/core/AudioManager.lua
local class = {}
class._bgmVolume = 100
class._bgsVolume = 100
class._meVolume = 100
class._seVolume = 100
class._currentBGM = nil -- 별도 객체
class._currentBGS = nil -- 별도 객체
class._bgm = nil
class._bgs = nil
class._me = nil
class._ses = {}
class._replayFadeTime = 0.5

local function updateCurrentBGM(self, bgm, pos)
    self._currentBGM = {file = bgm.file, volume = bgm.volume, pos = pos}
end
local function updateCurrentBGS(self, bgm, pos)
    self._currentBGS = {file = bgm.file, volume = bgm.volume, pos = pos}
end
local function cleanupSE(self)
    self._ses = Array.filter(self._ses, function(item)
        ret = item:isPlaying()
        if not ret then item:Unload() end
        return ret
    end)
end

function class:PlayBGM(bgm, pos)
    if self:isCurrentBGM(bgm) then
        self._bgm:Volume(self._bgmVolume)
    else
        self:StopBGM()
        if bgm ~= nil then
            self._bgm = bgm
            self._bgm:Volume(self._bgmVolume)
            if self._me == nil then
                self._bgm:Seek(pos or 0)
                self._bgm:Play()
            end
        end
    end
    updateCurrentBGM(self, bgm, pos)
end
function class:ReplayBGM(bgm)
    if self:isCurrentBGM(bgm) then
        self._bgm:Volume(self._bgmVolume)
    else
        self:PlayBGM(bgm, bgm:Pos())
        if bgm ~= nil then
            self._bgm = bgm
            if self._bgm ~= nil then
                self._bgm:FadeIn(self._replayFadeTime)
            end
        end
    end
end
function class:StopBGM()
    if self._bgm ~= nil then
        self._bgm:Unload()
        self._bgm = nil
        self._currentBGM = nil
    end
end
function class:FadeOutBGM(dur)
    if self._bgm ~= nil and self._currentBGM ~= nil then
        self._bgm:FadeOut(dur)
        self._currentBGM = nil
    end
end
function class:FadeInBGM(dur)
    if self._bgm ~= nil and self._currentBGM ~= nil then
        self._bgm:FadeIn(dur)
    end
end
function class:isCurrentBGM(bgm)
    return self._currentBGM ~= nil and self._bgm ~= nil and
               self._currentBGM.file == bgm.file
end

function class:PlayBGS(bgs, pos)
    if self:isCurrentBGS(bgs) then
        self._bgs:Volume(self._bgsVolume)
    else
        self:StopBGS()
        if bgs ~= nil then
            self._bgs = bgs
            self._bgs:Volume(self._bgsVolume)
            if self._me == nil then
                self._bgs:Seek(pos or 0)
                self._bgs:Play()
            end
        end
    end
    updateCurrentBGS(self, bgs, pos)
end
function class:ReplayBGS(bgs)
    if self:isCurrentBGS(bgs) then
        updateBGSParams(bgs)
    else
        self:PlayBGS(bgs, bgs:Pos())
        if bgs ~= nil then
            self._bgs = bgs
            if self._bgs ~= nil then
                self._bgs:FadeIn(self._replayFadeTime)
            end
        end
    end
end
function class:StopBGS()
    if self._bgs ~= nil then
        self._bgs:Unload()
        self._bgs = nil
        self._currentBGS = nil
    end
end
function class:FadeOutBGS(dur)
    if self._bgs ~= nil and self._currentBGS ~= nil then
        self._bgs:FadeOut(dur)
        self._currentBGS = nil
    end
end
function class:FadeInBGS(dur)
    if self._bgs ~= nil and self._currentBGS ~= nil then
        self._bgs:FadeIn(dur)
    end
end
function class:isCurrentBGS(bgs)
    return self._currentBGS ~= nil and self._bgs ~= nil and
               self._currentBGS.file == bgs.file
end

function class:PlayME(me)
    self:StopME()
    if me ~= nil then
        if self._bgm and self._currentBGM then
            self._currentBGM.pos = self._bgm:Pos()
            self._bgm:Pause()
        end

        self._me = me
        self._me:Volume(self._meVolume)
        self._me:Play()
    end
end
function class:FadeOutME(dur) if self._me ~= nil then self._me.FadeOut(dur) end end
function class:StopME()
    if self._me ~= nil then
        self._me:Unload()
        self._me = nil

        if self._bgm ~= nil and self._currentBGM ~= nil and
            not self._bgm:isPlaying() then
            self._bgm:Play()
            self._bgm:FadeIn(self._replayFadeTime)
        end
    end
end

function class:PlaySE(se)
    if se ~= nil then
        latest = Array.filter(self._ses, function(audio)
            return audio.frameCount == Game.frameCount and audio.file == se.file
        end)
        if #latest > 0 then return end -- prevent play same se at same frame

        se:Volume(self._seVolume)
        se.Play()
        Array.push(self._ses, se)
        cleanupSE(self)
    end
end
function class:StopSE()
    Array.foreach(self._ses, function(item)
        item:Stop()
        item:Unload()
    end)
    self._ses = {}
end

function class:StopAll()
    self:StopME()
    self:StopBGM()
    self:StopBGS()
    self:StopSE()
end

function class:SaveBGM()
    if self._bgm ~= nil then
        updateCurrentBGM(self, self._bgm, self._bgm:Pos())
        return self._currentBGM
    else
        return {file = "", volume = 0, pos = 0}
    end
end
function class:SaveBGS()
    if self._bgs ~= nil then
        updateCurrentBGS(self, self._bgs, self._bgs:Pos())
        return self._currentBGS
    else
        return {file = "", volume = 0, pos = 0}
    end
end

AudioManager = class

File: web/script/src/managers/AudioManager.ts
import Game from "..";
import WebAudio from "@/core/WebAudio";
import Global from "@/core/Global";

export interface AudioMetadata {
	name: string;
	volume: number;
	pos?: number;
	loop?: boolean;
}

interface MergedAudio {
	metadata: AudioMetadata;
	audio: WebAudio;
}

class AudioManagerClass {
	public static readonly audioFileExt = ".mp3";

	private _bgmVolume = Global.get<number>("volume.bgm", 100);
	public get bgmVolume () { return this._bgmVolume; }
	public set bgmVolume (value: number) {
		this._bgmVolume = value;
		this.updateBGMParams(this._currentBGM);
	}

	private _seVolume = Global.get<number>("volume.sfx", 100);
	public get seVolume () { return this._seVolume; }
	public set seVolume (value: number) {
		this._seVolume = value;
		this._seList.forEach(x => this.updateSEParams(x.audio, x.metadata));
	}

	private _blocked = false;
	public get blocked () { return this._blocked; }
	public set blocked (value: boolean) {
		this._blocked = value;

		if (this._currentBGM)
			this._bgmBuffer?.play(true);
	}

	private _currentBGM: AudioMetadata | null = null;
	private _bgmBuffer: WebAudio | null = null;
	private readonly _seList: MergedAudio[] = [];
	private _replayFadeTime = 0.5;
	private _path = "assets/";

	public playBGM (bgm: AudioMetadata, pos: number = 0) {
		if (this.isCurrentBGM(bgm))
			this.updateBGMParams(bgm);
		else {
			this.stopBGM();
			if (bgm.name) {
				this._bgmBuffer = this.createBuffer("BGM/", bgm.name);
				this.updateBGMParams(bgm);

				if (!this.blocked)
					this._bgmBuffer.play(true);
			}
		}
		this.updateCurrentBGM(bgm, pos);
	}
	public replayBGM (bgm: AudioMetadata) {
		if (this.isCurrentBGM(bgm))
			this.updateBGMParams(bgm);
		else {
			this.playBGM(bgm, bgm.pos);
			if (this._bgmBuffer)
				this._bgmBuffer.fadeIn(this._replayFadeTime);
		}
	}
	public isCurrentBGM (bgm: AudioMetadata) {
		return this._currentBGM && this._bgmBuffer && this._currentBGM.name === bgm.name;
	}
	private updateBGMParams (bgm: AudioMetadata | null) {
		this.updateBufferParams(this._bgmBuffer, this._bgmVolume, bgm);
	}
	private updateCurrentBGM (bgm: AudioMetadata, pos: number) {
		this._currentBGM = {
			name: bgm.name,
			volume: bgm.volume,
			pos: pos
		}
	}
	public stopBGM () {
		if (this._bgmBuffer) {
			this._bgmBuffer.destroy();
			this._bgmBuffer = null;
			this._currentBGM = null;
		}
	}
	public fadeOutBGM (duration: number) {
		if (this._bgmBuffer && this._currentBGM) {
			this._bgmBuffer.fadeOut(duration);
			this._currentBGM = null;
		}
	}
	public fadeInBGM (duration: number) {
		if (this._bgmBuffer && this._currentBGM) {
			this._bgmBuffer.fadeIn(duration);
		}
	}

	public playSE (se: AudioMetadata) {
		if (this.blocked) return;

		if (se.name) {
			// 같은 프레임에 같은 SE 동시 재생 방지
			if (this._seList.some(se => se.audio.frameCount === Game.frameCount && buffer.name === se.audio.name))
				return;

			const buffer = this.createBuffer("SE/", se.name);
			this.updateSEParams(buffer, se);
			buffer.play(se.loop || false);

			this.cleanupSE();
			this._seList.push({
				audio: buffer,
				metadata: se,
			});
		}
	}
	public updateSEParams (buffer: WebAudio, se: AudioMetadata) {
		this.updateBufferParams(buffer, this._seVolume, se);
	}
	public cleanupSE () {
		for (const buffer of this._seList)
			if (!buffer.audio.isPlaying())
				buffer.audio.destroy();

		let i = 0;
		while (i < this._seList.length) {
			if (this._seList[i].audio.isPlaying()) {
				i++;
				continue;
			}
			this._seList.splice(i, 1);
		}
	}
	public stopSE () {
		this._seList.forEach(se => se.audio.destroy());
		this._seList.splice(0, this._seList.length);
	}

	public stopAll () {
		this.stopBGM();
		this.stopSE();
	}

	public saveBGM (): AudioMetadata {
		if (this._currentBGM) {
			const bgm = this._currentBGM;
			return {
				name: bgm.name,
				volume: bgm.volume,
				pos: this._bgmBuffer ? this._bgmBuffer.time() : 0
			}
		} else {
			return this.makeEmptyAudioObject();
		}
	}
	public makeEmptyAudioObject (): AudioMetadata {
		return { name: "", volume: 0 }
	}

	public createBuffer (dir: string, name: string) {
		const url = this._path + dir + encodeURIComponent(name).replace(/%2F/, "/") + AudioManagerClass.audioFileExt;
		const buffer = new WebAudio(url);
		buffer.name = name;
		buffer.frameCount = Game.frameCount;
		return buffer;
	}

	public updateBufferParams (buffer: WebAudio | null, configVolume: number, audio: AudioMetadata | null) {
		if (buffer && audio)
			buffer.volume = (configVolume * (audio.volume || 0)) / 10000;
	}

	public checkErrors () {
		const buffers = [this._bgmBuffer, ...this._seList.map(x => x.audio)];
		for (const buffer of buffers) {
			if (buffer && buffer.isError())
				this.throwLoadError(buffer);
		}
	}
	public throwLoadError (webAudio: WebAudio) {
		const reload = webAudio.reload.bind(webAudio);
		throw ["LoadError", webAudio.url, reload];
	}
}

const AudioManager = new AudioManagerClass();
export default AudioManager;
